### DISCUSSION TITLE
How to mock external services?

### DISCUSSION THREAD

**Discussion by rdmolony** [seleniumbase/SeleniumBase#1142] [Q&A] [answered]
Hi there, I have a dodgy `streamlit` app that depends on a response from an external service.  I'm struggling to mock this service in my `Selenium` functional tests.  Trawling [`stackexchange`](https://sqa.stackexchange.com/questions/13668/mocking-http-reponses-in-selenium-web-driver-using-python) & the [`Selenium`](https://www.selenium.dev/documentation/webdriver/capabilities/shared/#proxy) docs it seems best practice is mocking this response using a proxy.

Is there a `SeleniumBase` best practice for this?

(SeleniumBase is great btw, thanks!)

---

The `stackexchange` answer recommends `browsermob-proxy-py` (**last commit 2018**):

```python
import browsermobproxy
import selenium.webdriver as driver
import time

website_homepage_url = "http://www.python.org"

server = browsermobproxy.Server("C:\Users\johnsmith\Computer_Code\\browsermob-proxy-2.1.0-beta-1\\bin\\browsermob-proxy")
server.start()
proxy = server.create_proxy()

profile = driver.FirefoxProfile()
profile.set_proxy(proxy.selenium_proxy())
browser = driver.Firefox(firefox_profile=profile)

proxy.new_har("Python.org")
proxy.rewrite_url(website_homepage_url, 'http://localhost:8000/Fake_Search_Results_Page.html')
browser.get(website_homepage_url)

time.sleep(10)

proxy.har  # returns a HAR JSON blob

server.stop()
browser.quit()
```

**Comment by mdmintz** [comment, marked as answer]
Hello @rdmolony, SeleniumBase allows you to set a proxy server when running tests, which you can use together with the browsermobproxy. There are two ways to set the proxy.
1: From the command-line with pytest:
``--proxy="server:port"``
2: By spinning up a new driver from within a test:
``driver2 = self.get_new_driver(proxy="server:port")``
If authentication is required, you can expand that format to ``user:pass@server:port``.
If using option 2, all ``self.*`` methods will automatically start using the new driver that you have spun up. Assuming that you have correctly set up your proxy server, the web browser will use that while tests are running.

**Reply by rdmolony** [reply]
Awesome, thanks a lot @mdmintz!  I saw your CLI flag in your wonderful docs & `get_new_driver` is intuitive ;) Do you have any preferred proxy library?

**Reply by mdmintz** [reply]
@rdmolony, Excellent. :) I don't have any preferred proxy libraries... I've generally connected to existing proxy servers as needed via the above options.

**Reply by rdmolony** [reply]
I've since found this library [wkeeling/selenium-wire](https://github.com/wkeeling/selenium-wire#intercepting-requests-and-responses) which does exactly what I was looking for
